src/loader/tsbgw_launcher.c

```diff
@@ -0,0 +1,511 @@
+ *
+ * Notes: multiple databases in an instance (PG cluster) can have Timescale installed. They are not necessarily
+ * the same version of Timescale (though they could be)
+ * Shared memory is allocated and background workers are registered at shared_preload_libraries time
+ * We do not know what databases exist, nor which databases Timescale is installed in (if any) at
+ * shared_preload_libraries time.
+ */
+
+
+
+
+typedef struct tsbgwHashEntry
+{
+	Oid			db_oid;			/* key for the hash table, must be first */
+	BackgroundWorkerHandle *db_scheduler_handle;	/* needed to shut down
+													 * properly */
+}			tsbgwHashEntry;
+
+
+
+
+
+
+
+/*
+ * Register a background worker that calls the main timescaledb library (ie loader) and uses the scheduler entrypoint function
+ * the scheduler entrypoint will deal with starting a new worker, and waiting on any txns that it needs to, if we pass along a vxid in the bgw_extra field of the BgWorker
+ *
+ */
+static bool
+register_tsbgw_entrypoint_for_db(Oid db_id, VirtualTransactionId vxid, BackgroundWorkerHandle **handle)
+{
+
+	BackgroundWorker worker;
+
+
+	memset(&worker, 0, sizeof(worker));
+	snprintf(worker.bgw_name, BGW_MAXLEN, "Timescale BGW Scheduler Entrypoint");
+	worker.bgw_flags = BGWORKER_SHMEM_ACCESS | BGWORKER_BACKEND_DATABASE_CONNECTION;
+	worker.bgw_restart_time = BGW_NEVER_RESTART;
+	worker.bgw_start_time = BgWorkerStart_RecoveryFinished;
+	sprintf(worker.bgw_library_name, EXTENSION_NAME);
+	sprintf(worker.bgw_function_name, TSBGW_ENTRYPOINT_FUNCNAME);
+	worker.bgw_notify_pid = MyProcPid;
+	worker.bgw_main_arg = db_id;
+	memcpy(worker.bgw_extra, &vxid, sizeof(VirtualTransactionId));
+
+	return RegisterDynamicBackgroundWorker(&worker, handle);
+
+}
+
+static volatile sig_atomic_t got_sigterm = false;
+static void
+tsbgw_sigterm(SIGNAL_ARGS)
+{
+	int			save_errno = errno;
+
+	got_sigterm = true;
+	SetLatch(MyLatch);
+
+	errno = save_errno;
+}
+
+
+
+/*
+ * Model this on autovacuum.c -> get_database_list
+ * Note that we are not doing all the things around memory context that they do, because
+ * a) we're using shared memory to store the list of dbs and b) we're in a function and
+ * shorter lived context here.
+ * This can get called at two different times 1) when the cluster launcher starts and is looking for dbs
+ * and 2) if it restarts due to a postmaster signal.
+ */
+static HTAB *
+populate_database_htab(void)
+{
+	Relation	rel;
+	HeapScanDesc scan;
+	HeapTuple	tup;
+	HTAB	   *db_htab = NULL;
+	HASHCTL		info;
+
+	/*
+	 * first initialize the hashtable before we start a txn, we'll be writing
+	 * info about dbs there
+	 */
+	info = (HASHCTL)
+	{
+		.keysize = sizeof(Oid),
+			.entrysize = sizeof(tsbgwHashEntry)
+	};
+	db_htab = hash_create("tsbgw_launcher_db_htab", 8, &info, HASH_BLOBS | HASH_ELEM);
+
+
+	/*
+	 * by this time we should already be connected to the db, and only have
+	 * access to shared catalogs
+	 */
+	/* start a txn, see note in autovacuum.c for why */
```

Comment:
This is the comment: 	/*
	 * Start a transaction so we can access pg_database, and get a snapshot.
	 * We don't have a use for the snapshot itself, but we're interested in
	 * the secondary effect that it sets RecentGlobalXmin.  (This is critical
	 * for anything that reads heap pages, because HOT may decide to prune
	 * them even if the process doesn't attempt to modify any tuples.)
	 */

I actually think this'n is a note we don't need, it's mainly there because in autovacuum we almost never work inside a txn. So it's notable that we are here, I'm just going to remove the comment as I would have thought that getting the db list requires a txn so we can see the correct things in pg_database.